### PR TITLE
[FEAT] Add Grafana feature gate

### PR DESCRIPTION
**Pull request by ritaCanavarro**
This PR solves the following issue: https://github.com/grafana/tempo-operator/issues/423

### CODE REVIEW

File: internal/manifests/operator/grafana/datasource.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package grafana
+
+import (
+	grafanav1 "github.com/grafana-operator/grafana-operator/v5/api/v1beta1"
+	configv1alpha1 "github.com/grafana/tempo-operator/apis/config/v1alpha1"
+)
+
+func Datasource(featureGates configv1alpha1.FeatureGates, namespace string) *grafanav1. {
```
Comment (frzifus) [inline internal/manifests/operator/grafana/datasource.go:8]: Is this file still used?

File: internal/manifests/operator/manifests.go [hunk 2]
```diff
@@ -5,8 +5,9 @@ import (
 
 	configv1alpha1 "github.com/grafana/tempo-operator/apis/config/v1alpha1"
 	"github.com/grafana/tempo-operator/internal/manifests/operator/prometheus"
-)
+	"go.searchlight.dev/grafana-operator"
```
Comment (frzifus) [inline internal/manifests/operator/manifests.go:8]: What is this all about?

**Review by frzifus — COMMENTED** [review]
Thanks @ritaCanavarro for picking this up. I see its in draft, but I added some comments.

File: internal/manifests/operator/grafana/datasource.go (hunk 1 above)
Comment (ritaCanavarro) [reply internal/manifests/operator/grafana/datasource.go:8]: I am going to work on this file, so yes.
This PR is still in draft so things might change.

File: internal/manifests/operator/manifests.go (hunk 2 above)
Comment (ritaCanavarro) [reply internal/manifests/operator/manifests.go]: Nevermind, I confused the packages. Going to remove this.

File: apis/tempo/v1alpha1/tempostack_types.go [hunk 3]
```diff
@@ -133,6 +133,13 @@ type ObservabilitySpec struct {
 	// +kubebuilder:validation:Optional
 	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Tracing Config"
 	Tracing TracingConfigSpec `json:"tracing,omitempty"`
+
+	// Datasources defines the datasources configuration for operands.
+	//
+	// +optional
+	// +kubebuilder:validation:Optional
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Datasources Config"
+	Datasources DatasourcesConfigSpec `json:"datasources,omitempty"`
```
Comment (andreasgerstmayr) [inline apis/tempo/v1alpha1/tempostack_types.go:142]: I'm wondering if we should rename this to `grafana` instead. Then we'll have `metrics`, `tracing` and `grafana` in `.spec.observability`. Wdyt @pavolloffay?

File: apis/tempo/v1alpha1/tempostack_types.go [hunk 4]
```diff
@@ -170,6 +177,23 @@ type TracingConfigSpec struct {
...
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Create CreateDatasource for Tempo"
+	CreateDatasource bool `json:"createServiceMonitors,omitempty"`
+
+	// JaegerAgentEndpoint defines the jaeger endpoint data gets send to.
+	//
+	// +optional
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:="localhost:8080"
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Tempo-Gateway-Endpoint"
+	TempoGatewayEndpoint string `json:"tempo_gateway_endpoint,omitempty"`
```
Comment (andreasgerstmayr) [inline apis/tempo/v1alpha1/tempostack_types.go:194]: The operator can figure out the endpoint. It's either query-frontend or the gateway (if enabled), for example
```
http://tempo-simplest-query-frontend.$namespace.svc:3200
```
The service name is generated here: https://github.com/grafana/tempo-operator/blob/a701112a9c1b3557b4c3b3d828c84629b33ca597/internal/manifests/queryfrontend/query_frontend.go#L342
The namespace of the service is in `tempo.Namespace` (where tempo is the `TempoStack` instance).

File: apis/config/v1alpha1/projectconfig_types.go [hunk 5]
```diff
@@ -97,10 +97,19 @@ type MetricsFeatureGates struct {
...
+	CreateDatasources bool `json:"createDatasources,omitempty"`
+}
+
 // ObservabilityFeatureGates configures observability of the operator.
 type ObservabilityFeatureGates struct {
 	// Metrics configures metrics of the operator.
 	Metrics MetricsFeatureGates `json:"metrics,omitempty"`
+
+	// Datasources configures datasources of the operator.
+	Datasources DatasourcesFeatureGates `json:"datasources,omitempty"`
```
Comment (andreasgerstmayr) [inline apis/config/v1alpha1/projectconfig_types.go:112]: imho we don't need this setting, as the user can enable/disable the creation of the datasource in each TempoStack instance.

File: config/overlays/openshift/controller_manager_config.yaml [hunk 6]
```diff
@@ -30,6 +30,7 @@ featureGates:
     openshiftRoute: true
     servingCertsService: true
   prometheusOperator: true
+  grafanaOperator: true
```
Comment (andreasgerstmayr) [inline config/overlays/openshift/controller_manager_config.yaml:33]:
```suggestion
  grafanaOperator: false
```

The operator is not installed by default on OpenShift.

File: controllers/tempo/operator_controller.go [hunk 7]
```diff
@@ -142,5 +143,21 @@ func (r *OperatorReconciler) pruneObjects(ctx context.Context, featureGates conf
 		}
 	}
 
+	if featureGates.GrafanaOperator {
```
Comment (andreasgerstmayr) [inline controllers/tempo/operator_controller.go:146]: `operator_controller.go` is only for objects of the operator itself ("global" things, not per TempoStack instance). Can you move/adapt this code to `findObjectsOwnedByTempoOperator` in `tempostack_create_or_update.go`? This function looks for all objects created by a specific TempoStack instance and prunes them if they are no longer required.

File: internal/manifests/operator/manifests.go [hunk 8]
```diff
@@ -24,5 +25,14 @@ func BuildAll(featureGates configv1alpha1.FeatureGates, namespace string) ([]cli
 		manifests = append(manifests, prometheusRule)
 	}
 
+	if featureGates.Observability.Datasources.CreateDatasources {
```
Comment (andreasgerstmayr) [inline internal/manifests/operator/manifests.go:28]: Can you move/adapt this code to https://github.com/grafana/tempo-operator/blob/a701112a9c1b3557b4c3b3d828c84629b33ca597/internal/manifests/manifests.go#L22?

`internal/manifests/operator` is only for objects for the operator itself, not for objects we create for each `TempoStack` instance.

**Review by andreasgerstmayr — COMMENTED** [review]
Looks excellent :100: 
I've left a few comments.

File: apis/tempo/v1alpha1/tempostack_types.go (hunk 3 above)
Comment (pavolloffay) [reply apis/tempo/v1alpha1/tempostack_types.go:142]: +1

File: apis/tempo/v1alpha1/tempostack_types.go (hunk 4 above)
Comment (ritaCanavarro) [reply apis/tempo/v1alpha1/tempostack_types.go]: I was unsure if I could use it or not. Will take that into consideration then, thanks.

File: apis/tempo/v1alpha1/tempostack_types.go (hunk 3 above)
Comment (ritaCanavarro) [reply apis/tempo/v1alpha1/tempostack_types.go]: Renamed it.

File: config/overlays/openshift/controller_manager_config.yaml (hunk 6 above)
Comment (ritaCanavarro) [reply config/overlays/openshift/controller_manager_config.yaml:33]: Will change it.

File: apis/config/v1alpha1/projectconfig_types.go (hunk 5 above)
Comment (ritaCanavarro) [reply apis/config/v1alpha1/projectconfig_types.go:112]: Yes, I can see it now. Will remove.

File: controllers/tempo/operator_controller.go (hunk 7 above)
Comment (ritaCanavarro) [reply controllers/tempo/operator_controller.go:146]: Thanks for the explanation, will move!

File: internal/manifests/operator/manifests.go (hunk 8 above)
Comment (ritaCanavarro) [reply internal/manifests/operator/manifests.go]: Thanks for the explanation, will move!

File: controllers/tempo/operator_controller.go (hunk 7 above)
Comment (ritaCanavarro) [reply controllers/tempo/operator_controller.go]: I have moved the code to the correct file, could you please validate it? :)

File: internal/manifests/operator/manifests.go (hunk 8 above)
Comment (ritaCanavarro) [reply internal/manifests/operator/manifests.go]: Moved and will start to work on remaining logic.

File: apis/tempo/v1alpha1/tempostack_types.go (hunk 4 above)
Comment (ritaCanavarro) [reply apis/tempo/v1alpha1/tempostack_types.go]: I have update the URL logic, could you please validate it?

File: apis/config/v1alpha1/projectconfig_validation.go [hunk 9]
```diff
@@ -42,6 +42,9 @@ func (c *ProjectConfig) Validate() error {
 	if c.Gates.Observability.Metrics.CreatePrometheusRules && !c.Gates.Observability.Metrics.CreateServiceMonitors {
 		return errors.New("the Prometheus rules alert based on collected metrics, therefore the createServiceMonitors feature must be enabled when enabling the createPrometheusRules feature")
 	}
+	if c.Gates.Observability.Datasources.CreateDatasources && c.Gates.GrafanaOperator {
```
Comment (andreasgerstmayr) [inline apis/config/v1alpha1/projectconfig_validation.go:45]: This validation is now obsolete and can be removed

File: apis/tempo/v1alpha1/tempostack_webhook.go [hunk 10]
```diff
@@ -333,6 +333,13 @@ func (v *validator) validateObservability(tempo TempoStack) field.ErrorList {
 		}
 	}
 
+	if tempo.Spec.Observability.Grafana.CreateDatasource && !v.ctrlConfig.Gates.GrafanaOperator {
+		return field.ErrorList{
+			field.Invalid(metricsBase.Child("createDatasource"), tempo.Spec.Observability.Grafana.CreateDatasource,
```
Comment (andreasgerstmayr) [inline apis/tempo/v1alpha1/tempostack_webhook.go:338]:
```suggestion
			field.Invalid(grafanaBase.Child("createDatasource"), tempo.Spec.Observability.Grafana.CreateDatasource,
```

Before the condition, declare
```
grafanaBase := observabilityBase.Child("grafana")
```

File: apis/tempo/v1alpha1/tempostack_types.go [hunk 11]
```diff
@@ -170,6 +177,15 @@ type TracingConfigSpec struct {
 	JaegerAgentEndpoint string `json:"jaeger_agent_endpoint,omitempty"`
 }
 
+type GrafanaConfigSpec struct {
+	// CreateDatasource specifies if a Datasource should be created for Tempo.
```
Comment (andreasgerstmayr) [inline apis/tempo/v1alpha1/tempostack_types.go:181]:
```suggestion
	// CreateDatasource specifies if a Grafana Datasource should be created for Tempo.
```

File: internal/manifests/datasource/grafana.go [hunk 12]
```diff
@@ -0,0 +1,38 @@
+package datasource
```
Comment (andreasgerstmayr) [inline internal/manifests/datasource/grafana.go:1]:
```suggestion
package grafana
```

Can we rename this package (and directory) to `grafana` and this file to `datasource.go`?

File: internal/manifests/datasource/grafana.go [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+			Name:      "tempo",
+			Namespace: namespace,
+		},
+		Spec: grafanav1.GrafanaDatasourceSpec{
+			DatasourceSpec: grafanav1.DatasourceSpec{
+				Access: "proxy",
+				Name:  "Tempo",
+				Type: "tempo",
+				// TODO: Change URL
+				URL:  featureGates.Observability.Datasources.TempoGatewayEndpoint,
```
Comment (andreasgerstmayr) [inline internal/manifests/datasource/grafana.go:28]:
```suggestion
				URL:  fmt.Sprintf("https://%s:%d", naming.ServiceFqdn(params.Tempo.Namespace, params.Tempo.Name, manifestutils.QueryFrontendComponentName), manifestutils.PortHTTPServer),
```

File: internal/manifests/datasource/grafana.go [hunk 14]
```diff
@@ -0,0 +1,38 @@
+package datasource
+
+import (
+	grafanav1 "github.com/grafana-operator/grafana-operator/v5/api/v1beta1"
+	configv1alpha1 "github.com/grafana/tempo-operator/apis/config/v1alpha1"
+	"github.com/grafana/tempo-operator/internal/manifests/manifestutils"
+)
+
+// Datasource creates a Datasource for Grafana Tempo.
+func Datasource(featureGates configv1alpha1.FeatureGates, namespace string) *grafanav1.GrafanaDatasource {
```
Comment (andreasgerstmayr) [inline internal/manifests/datasource/grafana.go:10]:
```suggestion
func Datasource(params manifestutils.Params) *grafanav1.GrafanaDatasource {
```
I'd suggest using `manifestutils.Params` here, to have access to the Tempo spec (containg the namespace) and all feature gates like with the other manifests, e.g. https://github.com/grafana/tempo-operator/blob/bb9d581a4be99b3d551f449c68374911a412e703/internal/manifests/distributor/distributor.go#L21

**Review by andreasgerstmayr — COMMENTED** [review]
Looks good! Almost good to merge, I've reviewed the current code and left a few comments.

File: apis/tempo/v1alpha1/tempostack_webhook.go (hunk 10 above)
Comment (ritaCanavarro) [reply apis/tempo/v1alpha1/tempostack_webhook.go]: Updated, I missed that 😶‍🌫️

File: apis/tempo/v1alpha1/tempostack_types.go (hunk 11 above)
Comment (ritaCanavarro) [reply apis/tempo/v1alpha1/tempostack_types.go]: Updated.

File: internal/manifests/datasource/grafana.go (hunk 14 above)
Comment (ritaCanavarro) [reply internal/manifests/datasource/grafana.go]: Agree, will change.

File: internal/manifests/datasource/grafana.go (hunk 13 above)
Comment (andreasgerstmayr) [reply internal/manifests/datasource/grafana.go]: Good point! Yes, we should.
I was thinking we support the basic setup now without the gateway (and get this PR merged) and then extend it with support for the gateway in a follow-up PR.

Comment (andreasgerstmayr) [reply internal/manifests/datasource/grafana.go]: Yep, sounds good!

**Comment by andreasgerstmayr** [comment]
The last thing we need before merging is a unit test. Can you add a test to `datasource_test.go`?
You can take https://github.com/grafana/tempo-operator/blob/bb9d581a4be99b3d551f449c68374911a412e703/internal/manifests/ingester/ingester_test.go as an example.

**Review by andreasgerstmayr — APPROVED** [review]
Thanks a lot for your contributions! :100:

**Pull request merged**